### PR TITLE
WIP warn if parsing with du_parse

### LINKED ISSUE #12585: Inconsistent date parsing of to_datetime
As this comes up regularly:
- [x] https://github.com/pydata/pandas/issues/12583
- [x] https://github.com/pydata/pandas/issues/12501
- [x] https://github.com/pydata/pandas/issues/7348
- [x] https://github.com/pandas-dev/pandas/issues/15075
- [x] #12611 deal with `yearfirst/datefirst` in `read_csv`
- [x] #12730, compat with dateutil 2.5.2 `dayfirst=True, yearfirst=True` fixes
- [x] https://github.com/pydata/pandas/issues/14301 raise if inconsistency in `read_csv` with `dayfirst/yearfirst`
- [x] performance https://github.com/pandas-dev/pandas/issues/25848
**The problem**

Depending on the string formats, you can get inconsistent parsing (meaning: different interpretation of parts of the strings as days vs months between different values in the passed list/array of strings). 
For example:

```
In [22]: pd.to_datetime(["31/12/2014", "10/03/2011"])
Out[22]: DatetimeIndex(['2014-12-31', '2011-10-03'])
```

where the first is parsed as 'dayfirst' and the second not. Above particular case can be solved by passing `dayfirst=True`, but in any case, the default behaviour of `to_datetime` results in inconsistent parsing here. 
Another typical case if where you do provide `dayfirst=True` but have a malformed date, then it will fall back to 'not dayfirst' (https://github.com/pydata/pandas/issues/3341).

**The reason**

The reason is the flexibility of `dateutil.parser.parse`. This is in the first place also a feature, as it lets you parse messy datetime colums. But when it comes to flexibility regarding `dayfirst`/`yearfirst`, this gives unexpected results. 
So dateutil will 'try' to use `dayfirst=True`, but when this does not give a valid date, it will ignore `dayfirst=True` and parse the string anyway:

```
In [3]: dateutil.parser.parse('1/12/2012', dayfirst=True)
Out[3]: datetime.datetime(2012, 12, 1, 0, 0)

In [4]: dateutil.parser.parse('1/13/2012', dayfirst=True)
Out[4]: datetime.datetime(2012, 1, 13, 0, 0)
```

The same is true for `dayfirst=False` (the default) as this will also try with 'day first' if the other way around failed. 
The issue for the lack of strictness of dayfirst is actually https://github.com/pydata/pandas/issues/3341.

**Possible solutions**
- In some cases passing `dayfirst=True` is sufficient (if you don't have messy or malformed dates).
- The typical answer is saying: "if you want to be sure to have strict and consistent parsing, provide a `format` argument. 
  We should probably stress this more in the docstring of `to_datetime` 

But, regardless of the two points above, fact is that the _default_ behaviour of `to_datetime` (without passing any arguments) can do inconsistent parsing and so give unexpected results. 
Possible ways to have better default behaviour:
- Have the possibility to specify that `dateutil` should be strict about `dayfirst`. I raised an issue for that here: https://github.com/dateutil/dateutil/issues/214
- Changing the default of `infer_datetime_format` to `True` would solve it in some cases (where the first datetime has the correct format you want), but certainly not all.
- The idea of raising a warning when inconsistent parsing happened, is AFAIK not possible as pandas does not know how dateutil parsed the dates. And always warning when no format is passed or could be inferred (so in practice, when dateutil is used), is maybe to drastically, as the flexible parsing is also a feature.

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #12585 (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
The issue is that in
```
        df = DataFrame(
            [[1, 2], [3, 4]],
            columns=date_range("1/1/2013", "1/2/2013"),
            index=["A", "B"],
        )
        df.index.name = name

        with tm.assert_produces_warning(warn):
            result = df.reset_index()
```
, resetting the index will try to insert an element into a datetime index. hence, it'll try to parse the new element as a date, and this'll trigger the new warning

There's currently no way of catching multiple warnings, so perhaps that should be addressed first

File: pandas/_libs/tslibs/parsing.pyx [hunk 1]
```diff
@@ -238,6 +238,17 @@ cdef inline bint does_string_look_like_time(str parse_string):
 
     return 0 <= hour <= 23 and 0 <= minute <= 59
 
+from pandas.util._exceptions import find_stack_level
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/parsing.pyx:241]: is there an option that avoids importing from outside libs here?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/parsing.pyx:241]: yup, could set it explicitly each time - I'll do that once I've sorted out the tests

File: pandas/tests/dtypes/test_generic.py [hunk 2]
```diff
@@ -18,9 +18,9 @@
 class TestABCClasses:
     tuples = [[1, 2, 2], ["red", "blue", "red"]]
     multi_index = pd.MultiIndex.from_arrays(tuples, names=("number", "color"))
-    datetime_index = pd.to_datetime(["2000/1/1", "2010/1/1"])
+    datetime_index = pd.to_datetime(["1/1/2000", "1/1/2010"])
     timedelta_index = pd.to_timedelta(np.arange(5), unit="s")
-    period_index = pd.period_range("2000/1/1", "2010/1/1/", freq="M")
+    period_index = pd.period_range("1/1/2000", "1/1/2010", freq="M")
```
Comment (MarcoGorelli) [inline pandas/tests/dtypes/test_generic.py:23]: `pd.to_datetime(["2000/1/1", "2010/1/1"])` doesn't raise, but `pd.period_range("2000/1/1", "2010/1/1/", freq="M")` does

don't know why, need to investigate

File: pandas/_libs/tslibs/parsing.pyx [hunk 3]
```diff
@@ -263,6 +263,18 @@ cdef inline bint does_string_look_like_time(str parse_string):
...
+
+
+def du_parse_with_warning(*args, **kwargs):
+    parsed = du_parse(*args, **kwargs)
+    warnings.warn(
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/parsing.pyx:271]: do we take a perf hit here?

**Comment by MarcoGorelli** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

still working on this

**Comment by MarcoGorelli** [comment]
closing, as from https://github.com/pandas-dev/pandas/issues/12585#issuecomment-1249343334 I don't think this is the right solution

**Pull request closed without merging**